src/backend/cdb/dispatcher/cdbdisp_async.c

```diff
@@ -680,11 +685,37 @@ checkDispatchResult(CdbDispatcherState *ds, int timeout_sec)
 			if (timeout_sec >= 0 && diff_us >= timeout_sec * 1000000L)
 				break;
 		}
+		else if (should_cancel)
+		{
+			/* Make QEs come to their senses. */
+			signalQEs(pParms);
+
+			for (i = 0; i < db_count; i++)
+			{
+				dispatchResult = pParms->dispatchResultPtrArray[i];
+
+				/*
+				 * If we're cancelling the transaction due to an OOM, there
+				 * might not be enough memory to discard the result properly.
+				 * Try to free the async results first and pray.
+				 */
+				pqClearAsyncResult(dispatchResult->segdbDesc->conn);
+				cdbconn_discardResults(dispatchResult->segdbDesc, 20);
```

Comment:
There are many additional checks in the `cleanupQE` function, which already uses the `cdbconn_discardResults` function. Do we need to do these or similar checks here?